extend/resource/import: Provide more complete information about implementation, recommendations, and advanced features (#848)
hashicorp/terraform-website: content/source/docs/extend/resources/import.html.md

# Resources - Import

Adding import support for Terraform resources will allow existing infrastructure to be managed within Terraform. This type of enhancement generally requires a small to moderate amount of code changes.

~> **Note:** Operators are responsible for writing the appropriate configuration that will be associated with the resource import. This restriction may be removed in a future version of Terraform.

> When importing, the operator will specify the Terraform configuration address for the resource they wish to import, along with an identifier for import. The import identifier may be different than the resource identifier (`ResourceData.SetId()`) for compatibility reasons outlined [#importer-state-function] in the below.

Reviewer: ```suggestion
When importing, the operator will specify the Terraform configuration address for the resource they wish to import, along with an identifier for import. The import identifier may be different than the resource identifier (`ResourceData.SetId()`) for compatibility reasons outlined below in the [Importer State Function](#importer-state-function) section.
```

Markdown syntax.

```console
$ terraform import example_thing.foo abc123
```

## Overview of Implementation

> 

Reviewer: ```suggestion

Implementing import support requires three changes: an `Importer` `State` function in the resource code, a `TestStep` with `ImportState: true` in the acceptance tests, and documentation of the import ID format.

```

### Resource Code Implementation

> In the resource code (e.g. `resource_example_thing.go`), an implementation of `Importer` `State` function:

Reviewer: ```suggestion
In the resource code (e.g. `resource_example_thing.go`), implement an `Importer` `State` function:
```

```go
func resourceExampleThing() *schema.Resource {
  return &schema.Resource{
    /* ... existing Resource functions ... */
    Importer: &schema.ResourceImporter{
      State: /* ... */,
    },
  }
}
```

### Resource Acceptance Testing Implementation

> In the resource acceptance testing (e.g. `resource_example_thing_test.go`), an implementation of `TestStep`s with `ImportState: true`:

Reviewer: ```suggestion
In the resource acceptance testing (e.g. `resource_example_thing_test.go`), implement `TestStep`s with `ImportState: true`:
```

```go
func TestAccExampleThing_basic(t *testing.T) {
  /* ... potentially existing acceptance testing logic ... */

  resource.ParallelTest(t, resource.TestCase{
    /* ... existing TestCase functions ... */
    Steps: []resource.TestStep{
      /* ... existing TestStep ... */
      {
        ResourceName:      "example_thing.test",
        ImportState:       true,
        ImportStateVerify: true,
      },
    },
  })
}
```

### Resource Documentation Implementation

> In the resource documentation (e.g. `website/docs/r/example_thing.html.markdown`), the addition of `Import` documentation section at the bottom of the page:

Reviewer: ```suggestion
In the resource documentation (e.g. `website/docs/r/example_thing.html.markdown`), add an `Import` documentation section at the bottom of the page:
```
